### PR TITLE
add Hungarian solver from optax

### PR DESCRIPTION
**Pull request by marcocuturi**
Take advantage of https://github.com/google-deepmind/optax/pull/1083 to include unregularized OT solver for point clouds of equal size.

### CODE REVIEW

File: src/ott/geometry/costs.py [hunk 1]
```diff
@@ -347,6 +347,34 @@ def tree_unflatten(cls, aux_data, children):  # noqa: D102
     return cls(*aux_data)
 
 
+@jtu.register_pytree_node_class
+class EuclideanP(TICost):
```
Comment (michalk8) [inline src/ott/geometry/costs.py:351]: Please add to the docs.

File: src/ott/tools/unreg.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    geom: geometry object
+    paired_indices: Array of shape ``[2, n]``, of :math:`n` pairs
+      of indices, for which the optimal transport assigns mass. Namely, for each
+      index :math:`0 <= k < n`, if one has
+      :math:`i := \text{paired_indices}[0, k]` and
+      :math:`j := \text{paired_indices}[1, k]`, then point :math:`i` in
+      the first geometry sends mass to point :math:`j` in the second.
+  """
+  geom: geometry.Geometry
+  paired_indices: Optional[jnp.ndarray] = None
```
Comment (michalk8) [inline src/ott/tools/unreg.py:39]: Would remove the optional and default none.

File: src/ott/tools/unreg.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class HungarianOutput(NamedTuple):
+  r"""Output of the Hungarian solver.
+
+  Args:
+    geom: geometry object
+    paired_indices: Array of shape ``[2, n]``, of :math:`n` pairs
+      of indices, for which the optimal transport assigns mass. Namely, for each
+      index :math:`0 <= k < n`, if one has
```
Comment (michalk8) [inline src/ott/tools/unreg.py:33]: Use `\leq`

File: src/ott/tools/unreg.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  n, m = geom.shape
+  assert n == m, f"Hungarian can only match same # of points, got {n} and {m}."
+  i, j = assignment.hungarian_algorithm(geom.cost_matrix)
+
+  hungarian_out = HungarianOutput(geom=geom, paired_indices=jnp.stack((i, j)))
+  return jnp.sum(geom.cost_matrix[i, j]) / n, hungarian_out
+
+
+def wassdis_p(x: jnp.ndarray, y: jnp.ndarray, p: float = 2.0) -> float:
+  """Convenience wrapper on `hungarian` to get :term:`Wasserstein distance`.
```
Comment (michalk8) [inline src/ott/tools/unreg.py:69]: Would remove the `convenience wrapper` part and rephrase as
```
Compute the :term:`Wasserstein distance` using the Hungarian algoritm.
```
or similar.

File: src/ott/tools/unreg.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  def matrix(self) -> jesp.BCOO:
+    """``[n, n]`` transport matrix  in sparse format, with ``n`` NNZ entries."""
+    n, _ = self.geom.shape
+    unit_mass = jnp.ones((n,)) / n
+    indices = self.paired_indices.swapaxes(0, 1)
+    return jesp.BCOO((unit_mass, indices), shape=(n, n))
+
+
+def hungarian(geom: geometry.Geometry) -> Tuple[jnp.ndarray, HungarianOutput]:
+  """Solve assignment problem using Hungarian as implemented in optax.
```
Comment (michalk8) [inline src/ott/tools/unreg.py:51]: :mod:\`optax\`

File: src/ott/tools/unreg.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+from ott.geometry import costs, geometry, pointcloud
+
+__all__ = ["hungarian"]
+
+
+class HungarianOutput(NamedTuple):
+  r"""Output of the Hungarian solver.
+
+  Args:
+    geom: geometry object
```
Comment (michalk8) [inline src/ott/tools/unreg.py:30]: Geometry object.

File: src/ott/tools/unreg.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+from typing import NamedTuple, Optional, Tuple
+
+import jax.experimental.sparse as jesp
+import jax.numpy as jnp
+
+from optax import assignment
+
+from ott.geometry import costs, geometry, pointcloud
+
+__all__ = ["hungarian"]
```
Comment (michalk8) [inline src/ott/tools/unreg.py:23]: __all__ = ["HungarianOutput", "hungarian", "wassdis_p"]

File: tests/tools/unreg_test.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from ott.geometry import costs, pointcloud
+from ott.solvers import linear
+from ott.tools import unreg
+
+
+class TestHungarian:
+
+  @pytest.mark.parametrize("cost_fn", [costs.PNormP(1.3), None])
+  def test_matches_sink(self, rng: jax.Array, cost_fn: Optional[costs.CostFn]):
+    n, m, dim = 12, 12, 5
```
Comment (michalk8) [inline tests/tools/unreg_test.py:31]: Would remove `m`.

File: tests/tools/unreg_test.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    np.testing.assert_allclose(
+        out_sink.primal_cost, cost_hung, rtol=1e-3, atol=1e-3
+    )
+    np.testing.assert_allclose(
+        out_sink.matrix, out_hung.matrix.todense(), rtol=1e-3, atol=1e-3
+    )
+
+  @pytest.mark.parametrize("p", [1.3, 2.3])
+  def test_wass(self, rng: jax.Array, p: float):
+    n, m, dim = 12, 12, 5
```
Comment (michalk8) [inline tests/tools/unreg_test.py:46]: Would remove `m`

File: tests/tools/unreg_test.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    x, y = gen_data(rng1, n, m, dim)
+    geom = pointcloud.PointCloud(x, y, cost_fn=costs.EuclideanP(p=p))
+    cost_hung, _ = unreg.hungarian(geom)
+    w_p = unreg.wassdis_p(x, y, p)
+    np.testing.assert_allclose(w_p, cost_hung ** 1. / p, rtol=1e-3, atol=1e-3)
+
+
+def gen_data(rng: jax.Array, n: int, m: int,
+             dim: int) -> Tuple[jnp.ndarray, jnp.ndarray]:
+  rngs = jax.random.split(rng, 4)
```
Comment (michalk8) [inline tests/tools/unreg_test.py:57]: Split into 2 keys please.

File: tests/tools/unreg_test.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+class TestHungarian:
+
+  @pytest.mark.parametrize("cost_fn", [costs.PNormP(1.3), None])
+  def test_matches_sink(self, rng: jax.Array, cost_fn: Optional[costs.CostFn]):
+    n, m, dim = 12, 12, 5
+    rng1, rng2 = jax.random.split(rng, 2)
+    x, y = gen_data(rng1, n, m, dim)
+    geom = pointcloud.PointCloud(x, y, cost_fn=cost_fn, epsilon=.0005)
+    cost_hung, out_hung = unreg.hungarian(geom)
```
Comment (michalk8) [inline tests/tools/unreg_test.py:35]: Please add `jax.jit`.

File: tests/tools/unreg_test.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    )
+
+  @pytest.mark.parametrize("p", [1.3, 2.3])
+  def test_wass(self, rng: jax.Array, p: float):
+    n, m, dim = 12, 12, 5
+    rng1, rng2 = jax.random.split(rng, 2)
+    x, y = gen_data(rng1, n, m, dim)
+    geom = pointcloud.PointCloud(x, y, cost_fn=costs.EuclideanP(p=p))
+    cost_hung, _ = unreg.hungarian(geom)
+    w_p = unreg.wassdis_p(x, y, p)
```
Comment (michalk8) [inline tests/tools/unreg_test.py:51]: Please add `jax.jit`.

File: src/ott/tools/unreg.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  i, j = assignment.hungarian_algorithm(geom.cost_matrix)
+
+  hungarian_out = HungarianOutput(geom=geom, paired_indices=jnp.stack((i, j)))
+  return jnp.sum(geom.cost_matrix[i, j]) / n, hungarian_out
+
+
+def wassdis_p(x: jnp.ndarray, y: jnp.ndarray, p: float = 2.0) -> float:
+  """Convenience wrapper on `hungarian` to get :term:`Wasserstein distance`.
+
+  Uses :func:`~ott.tools.unreg.hungarian` to solve the
```
Comment (michalk8) [inline src/ott/tools/unreg.py:71]: Use :func:\`hungarian\`.

File: src/ott/tools/unreg.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    unit_mass = jnp.ones((n,)) / n
+    indices = self.paired_indices.swapaxes(0, 1)
+    return jesp.BCOO((unit_mass, indices), shape=(n, n))
+
+
+def hungarian(geom: geometry.Geometry) -> Tuple[jnp.ndarray, HungarianOutput]:
+  """Solve assignment problem using Hungarian as implemented in optax.
+
+  Args:
+    geom: (square) geometry object.
```
Comment (michalk8) [inline src/ott/tools/unreg.py:54]: Would say 
```
Geometry of shape ``[n, n]``.
```

File: src/ott/tools/unreg.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+import jax.numpy as jnp
+
+from optax import assignment
+
+from ott.geometry import costs, geometry, pointcloud
+
+__all__ = ["hungarian"]
+
+
+class HungarianOutput(NamedTuple):
```
Comment (michalk8) [inline src/ott/tools/unreg.py:26]: Would prefer to use a dataclass.

File: src/ott/tools/unreg.py (hunk 5 above)
Comment (michalk8) [reply src/ott/tools/unreg.py:51]: Also please add this to `intersphinx_mapping` in `conf.py`

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/geometry/costs.py (hunk 1 above)
Comment (marcocuturi) [reply src/ott/geometry/costs.py:351]: ok. Don't want to surprise the user because there's no Legendre transform, but can add.

File: src/ott/tools/unreg.py (hunk 5 above)
Comment (marcocuturi) [reply src/ott/tools/unreg.py:51]: isn't it there already? 
```
intersphinx_mapping = {
    "python": ("https://docs.python.org/3", None),
    "numpy": ("https://numpy.org/doc/stable/", None),
    "jax": ("https://jax.readthedocs.io/en/latest/", None),
    "jaxopt": ("https://jaxopt.github.io/stable", None),
    "lineax": ("https://docs.kidger.site/lineax/", None),
    "flax": ("https://flax.readthedocs.io/en/latest/", None),
    "optax": ("https://optax.readthedocs.io/en/latest/", None),
    "diffrax": ("https://docs.kidger.site/diffrax/", None),
    "scipy": ("https://docs.scipy.org/doc/scipy/", None),
    "pot": ("https://pythonot.github.io/", None),
    "matplotlib": ("https://matplotlib.org/stable/", None),
}
```

File: src/ott/tools/unreg.py (hunk 15 above)
Comment (marcocuturi) [reply src/ott/tools/unreg.py:26]: all our output objects are `NamedTuple` at the moment, i feel this should be done with a refactoring and not a local change, no?
